### PR TITLE
[Ready] Anomaly Refinement Overhaul

### PR DESCRIPTION
**Pull request by BurgerLUA**
## About The Pull Request

Overhauls requirements for Anomaly Refinement, and removes some limits and implements some more.

The Anomaly Refiner now:
- Has new sounds to indicate stages of the refinement process.
- Requires 1.21 Megawatts of power per core implosion of the machine.
- Takes 1 minute to refine a core.
- No longer shakes the station or makes a loud noise.
- Has a minimum requirement of 25% of the bomb cap to successfully refine, which increases with each core used. This increasement diminishes over time.
- Has unlimited uses.

## Why It's Good For The Game

The station shaking for 5 minutes straight while the scientist refines cores all in one go was excessively annoying.  The limitation for anomaly cores was also arbitrary, especially considering that a lot of objects require cores and on some rounds you get a lot of anomaly spawns while others you don't.

This PR should fix all of the above and add some extra depth to the mechanics of the anomaly refinement process so it isn't just "make 20 maxcaps in one sitting then use all of them to make cores while the station complains about excessive explosions"

## Proof Of Testing

I tested it. It just werks :^)

## Changelog

:cl: BurgerBB
balance: Overhauls requirements for Anomaly Refinement, and removes some limits and implements some more.
/:cl:

### CODE REVIEW

**Comment by pixelkitty286** [comment]
After experiencing those rounds I feel like this is a salt pr what's next removing gibtonite from the game because of ice box?

**Comment by BurgerLUA** [comment]
> After experiencing those rounds I feel like this is a salt pr what's next removing gibtonite from the game because of ice box?

yeah

**Comment by AllCorruptingWyrm** [comment]
personally my screen shaking for 5m straight was annoying, not sure why this is being contested.

**Comment by codehonky** [comment]
Ordnance is already a grindy process to setup for refining cores - adding on an additional two minute wait and excessive energy pull per core is annoying and entirely unnecessary. There's already a wait time and financial cost for ordering the raw cores from cargo - did you take this into consideration? I've never once run into the anomaly refiner's "limit" in regard to uses either. 

The station shaking for a short while if a scientist decides to refine a bunch of cores at once really isn't that big a nuisance. Deal with it. If it's happening every round because a specific scientist or RD always refines a bunch of cores at once that don't even end up getting used - maybe talk to that individual instead and ask they don't do that round after round rather than making a whole PR.

Anomaly refinement is perfectly satisfactory as it exists and this PR shouldn't be merged.

**Comment by R3dtail** [comment]
Yeah I'm not gona lie making the screen not shake?  WOuld be great!  Considering I'm the only person who does mass refining,  I feel a little bad basic ally making everyone do nothing till i'm done.  But uh, the two minute wait coupled with the EXTREME power  drain basically says "yeah no stop refining cores".  I don't like that.  It's hard enough to get engineering to put more power into the grid already and uh.  Two minutes is relatively a long time.  I don't see how this improves the experience for people directly doing ordnance.  Even with the limits as is you can make  WAY more than enough cores for everyone.  If this was just "stops the screen shaking", that'd be fine.  The set up process for ordnance is fun enough as is, everything else in this  PR I can already tell would make it annoying and tedious.

**Comment by projectkepler-ru** [comment]
@codehonky  I disagree on the financial cost, the RD has a shit ton of money to spend, and a lot of the time they're ordering it en mass. Which is what this PR is targeted toward

**Comment by R3dtail** [comment]
> @codehonky I disagree on the financial cost, the RD has a shit ton of money to spend, and a lot of the time they're ordering it en mass. Which is what this PR is targeted toward

Technically,  at least when I'm RD, the mass amount of money comes from doing ordnance experiments.  The anomaly refining is a victory lap for me.

**Comment by R3dtail** [comment]
Aaaaaalso someone brought up a good point.  Wouldn't the 1.5 MW draw just drive more of a point of contention between science and the rest of the station?  That's a HUGE amount of draw.  Engineering already tends to get frustrated with science for having a cell charger (NOT a multicell charger, that's universally frowned upon) running, let alone something that consistently draws 1.5 MW.  This further dissentivises core refining at all especially  with the 2 minute wait time.  This really does feel like soft removing core refining.

**Comment by codehonky** [comment]
> @codehonky I disagree on the financial cost, the RD has a shit ton of money to spend, and a lot of the time they're ordering it en mass. Which is what this PR is targeted toward

And? It's not as if R&D has an infinite amount of money and the budget doesn't magically appear out of nowhere. This PR is targeted toward curbing the behavior of a singular RD player - it has absolutely zero merit.

**Comment by pixelkitty286** [comment]
Simple solution here another server did it just remove explosions shaking the screen because another server got fed up with this too then proceeded to remove it.

**Comment by BurgerLUA** [comment]
Reduced the time taken to refine a core from 2 minutes to 1 minute.

**Comment by whatisthisplace** [comment]
So I have many questions:

1. What happens if the refiner runs out of power during refinement?
2. Whats going to happen when we have newbie engineers who cant output excess of 1MW in the grid?
3. Is there really a point in making a 1 minute powersink because someone wants flashy gear?

Why not just go the middle ground and just have the refinery not shake the station, instead of adding artificial timers, seems like the most logical step imo.

**Comment by Doom-external** [comment]
> No longer shakes the station or makes a loud noise.

This is all that was really needed, everything else is fine.

**Comment by projectkepler-ru** [comment]
> 
> And? It's not as if R&D has an infinite amount of money and the budget doesn't magically appear out of nowhere. This PR is targeted toward curbing the behavior of a singular RD player - it has absolutely zero merit.

They, do? they're the only department to start with 10k and easily reaches up to 60k as the round progress

**Comment by R3dtail** [comment]
> > And? It's not as if R&D has an infinite amount of money and the budget doesn't magically appear out of nowhere. This PR is targeted toward curbing the behavior of a singular RD player - it has absolutely zero merit.
> 
> They, do? they're the only department to start with 10k and easily reaches up to 60k as the round progress

It's more so they have the most direct tools of acquiring budget compared to other departments outside of Cargo.  Considering Ordnance is all about budget creation.  So it isn't so much that it appears out of nowhere, its because someone's toiled away in ordnance doing the relevant experiments to GET that funding.

**Comment by codehonky** [comment]
> > And? It's not as if R&D has an infinite amount of money and the budget doesn't magically appear out of nowhere. This PR is targeted toward curbing the behavior of a singular RD player - it has absolutely zero merit.
> 
> They, do? they're the only department to start with 10k and easily reaches up to 60k as the round progress

That 10k typically goes toward a tritium canister which helps later reach that 60k. Anomaly cores are ~2500 credits each and aren't the only things that science buys.

**Comment by projectkepler-ru** [comment]
> > > And? It's not as if R&D has an infinite amount of money and the budget doesn't magically appear out of nowhere. This PR is targeted toward curbing the behavior of a singular RD player - it has absolutely zero merit.
> > 
> > 
> > They, do? they're the only department to start with 10k and easily reaches up to 60k as the round progress
> 
> That 10k typically goes toward a tritium canister which helps later reach that 60k. Anomaly cores are ~2500 credits each and aren't the only things that science buys.

the tritrium canister is typically actually purchased with the console order

**Comment by codehonky** [comment]
> > > > And? It's not as if R&D has an infinite amount of money and the budget doesn't magically appear out of nowhere. This PR is targeted toward curbing the behavior of a singular RD player - it has absolutely zero merit.
> > > 
> > > 
> > > They, do? they're the only department to start with 10k and easily reaches up to 60k as the round progress
> > 
> > 
> > That 10k typically goes toward a tritium canister which helps later reach that 60k. Anomaly cores are ~2500 credits each and aren't the only things that science buys.
> 
> the tritrium canister is typically actually purchased with the console order

The console doesn't use up the department budget? Either way, the financial cost of anomaly cores is still a relevant factor in obtaining them. If nobody does the ordnance experiments, you can only order a few with other purchases considered.

None of this has to do with adding unreasonable wait times and energy costs to anomaly refinement all because one RD player refines a ton of cores at once every round they play.

**Comment by R3dtail** [comment]
I still don't quite get why there needed to be all these extra changes.  Just the screenshake being removed is such a good change, but the rest of these changes make doing any core refining such a massive burden that it's not worth it, period.

**Comment by BurgerLUA** [comment]
Reduced the time taken to refine a core from 1 minute to ~~5 years~~ 5 seconds.

**Comment by R3dtail** [comment]
5 seconds is definitely more bearable.  What about the power draw?  I get it, you want core refining to really mean something and be a thoughtful decision, but... 1.21 MW is a LOT of power.   Could we consider maybe 200 KW and go from there?  Please consider that  making  science even MORE of an extreme burden on the powergrid is just going to drive conflict that isn't fun for anyone to deal with between engineering and science.  The former of which is already a very stressful job and gets yelled at enough.

**Review by Majkl-J — CHANGES_REQUESTED** [review]
This is taking a simple issue of anomaly refinements being capped and shaking the station, and instead of just fixing those two, adding more issues by introducing arbitrary timers and comically large power requirements for the department that's already responsible for a shit ton of power drain.

Just fixing the main issue with the station loudly breaking it down each time a core is refined is more than enough

**Comment by BurgerLUA** [comment]
Refinement time is 5 seconds (this is what it is before this PR).
Power draw for each implosion is 0.25 MW.

File: modular_zubbers/modules/alt_anomaly_refinery/anomaly_refinery.dm [hunk 1]
```diff
@@ -0,0 +1,127 @@
+#define REFINERY_ANOMALY_REFINEMENT_TIME (5 SECONDS)
+#define REFINERY_ANOMALY_POWER_REQUIREMENT (0.25 MEGA WATTS)
```
Comment (Majkl-J) [inline modular_zubbers/modules/alt_anomaly_refinery/anomaly_refinery.dm:2]: Missing undefs

File: modular_zubbers/modules/alt_anomaly_refinery/anomaly_refinery.dm [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+	. += span_notice("Each core refinement takes <b>[DisplayTimeText(REFINERY_ANOMALY_REFINEMENT_TIME,1)]</b> to complete.")
+
+/obj/machinery/research/anomaly_refinery/get_required_radius(anomaly_type)
+
+	var/explosion_range = max(GLOB.MAX_EX_DEVESTATION_RANGE,GLOB.MAX_EX_HEAVY_RANGE,GLOB.MAX_EX_LIGHT_RANGE,0)
+	var/explosion_mod = 0.25 + (requirement_mod/100)*0.75
+
+	return round(explosion_range * explosion_mod,1)
+
+
```
Comment (Majkl-J) [inline modular_zubbers/modules/alt_anomaly_refinery/anomaly_refinery.dm:23]:
```suggestion
	var/explosion_range = max(GLOB.MAX_EX_DEVESTATION_RANGE, GLOB.MAX_EX_HEAVY_RANGE, GLOB.MAX_EX_LIGHT_RANGE, 0)
	var/explosion_mod = 0.25 + (requirement_mod/100)*0.75

	return round(explosion_range * explosion_mod,1)

```

**Review by Majkl-J — COMMENTED** [review]
Seems TM ready, I wanna give this a shot. A bit iffy on the whitespace usage, but looks okay otherwise.

Also no idea if this would be cleaner if it was partly non-modular incase TG updates this on their end. We'll see how it goes.

**Review by Majkl-J — CHANGES_REQUESTED** [review]
Also the sounds need to be actual explosion sounds this just sounds silly currently

**Comment by BurgerLUA** [comment]
> Also the sounds need to be actual explosion sounds this just sounds silly currently

Given that the shaking was suppressed by the machine inside, it makes no sense that sound would reach out of it. The minesweeper explosion sounds were added as a means of feedback.

**Comment by Majkl-J** [comment]
> Given that the shaking was suppressed by the machine inside, it makes no sense that sound would reach out of it. The minesweeper explosion sounds were added as a means of feedback.

It does not give the same feeling as even the distant explosion sound.

**Comment by nevimer** [comment]
needs more feedback, im uncertain

**Comment by BurgerLUA** [comment]
> needs more feedback, im uncertain

This hasn't been testmerged yet. A testmerge can help with setting a fair power value.

**Comment by StrangeWeirdKitten** [comment]
Due to recent power draw concerns being fixed upstream with how cells are handled, I have zero issues with giving this a shot. Anomaly gear is still fairly powerful per their design and making it require a little more work isn't really a concern of mine.

**Comment by R3dtail** [comment]
After experiencing it personally it doesn't feel too terrible.  Power's a bit more forgiving after recent changes,     Power i'm still a little iffy on, but 1.21 MW (NOW) after recent power changes is at least sustainable.  The sound effect used I find pleasant, too lol.

File: modular_zubbers/code/modules/alt_anomaly_refinery/anomaly_refinery.dm [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+		src,
+		pick('sound/effects/explosion1.ogg','sound/effects/explosion2.ogg','sound/effects/explosion3.ogg'),
+		50
+	)
+
+	return
+
+/obj/machinery/research/anomaly_refinery/proc/remove_requirement()
+
+	requirement_mod = max(0,requirement_mod-1)
```
Comment (Majkl-J) [inline modular_zubbers/code/modules/alt_anomaly_refinery/anomaly_refinery.dm:62]: ditto whitespace

```suggestion
/obj/machinery/research/anomaly_refinery/proc/remove_requirement()
	requirement_mod = max(0,requirement_mod-1)
```

File: modular_zubbers/code/modules/alt_anomaly_refinery/anomaly_refinery.dm [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+	if(!inserted_core)
+		test_status = "ERROR: No core present during detonation."
+		playsound(
+			src,
+			pick('modular_zubbers/sound/arcade/minesweeper_winfail.ogg'),
+			50
+		)
+		return COMSIG_CANCEL_EXPLOSION
+
```
Comment (Majkl-J) [inline modular_zubbers/code/modules/alt_anomaly_refinery/anomaly_refinery.dm:84]: Might be easier to just do . = COMSIG_CANCEL_EXPLOSION at the start of the proc and use empty returns

```suggestion
		)
		return

```

File: modular_zubbers/code/modules/alt_anomaly_refinery/anomaly_refinery.dm [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+		playsound(
+			src,
+			pick('modular_zubbers/sound/arcade/minesweeper_boardpress.ogg'),
+			50
+		)
+		say("Anomaly stability now at 100%.")
+
+/obj/machinery/research/anomaly_refinery/check_test(atom/source, list/arguments)
+
+	if(!inserted_core)
```
Comment (Majkl-J) [inline modular_zubbers/code/modules/alt_anomaly_refinery/anomaly_refinery.dm:76]: Ditto return change
```suggestion
/obj/machinery/research/anomaly_refinery/check_test(atom/source, list/arguments)
	. = COMSIG_CANCEL_EXPLOSION // We always want to cancel

	if(!inserted_core)
```

File: modular_zubbers/code/modules/alt_anomaly_refinery/anomaly_refinery.dm [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+	var/required_range = get_required_radius(inserted_core.anomaly_type)
+
+	if(explosion_range < required_range)
+		test_status = "Resultant detonation failed to produce enough implosive power to compress [inserted_core]. Items ejected."
+		playsound(
+			src,
+			pick('modular_zubbers/sound/arcade/minesweeper_winfail.ogg'),
+			50
+		)
+		return COMSIG_CANCEL_EXPLOSION
```
Comment (Majkl-J) [inline modular_zubbers/code/modules/alt_anomaly_refinery/anomaly_refinery.dm:98]: Ditto return change
```suggestion
		)
		return
```

File: modular_zubbers/code/modules/alt_anomaly_refinery/anomaly_refinery.dm [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+	if(test_status)
+		playsound(
+			src,
+			pick('modular_zubbers/sound/arcade/minesweeper_winfail.ogg'),
+			50
+		)
+		return COMSIG_CANCEL_EXPLOSION
```
Comment (Majkl-J) [inline modular_zubbers/code/modules/alt_anomaly_refinery/anomaly_refinery.dm:106]: Ditto return change
```suggestion
		)
		return
```

File: modular_zubbers/code/modules/alt_anomaly_refinery/anomaly_refinery.dm [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+	inserted_core = inserted_core.create_core(src, TRUE, TRUE)
+	test_status = "Success. Resultant detonation has theoretical range of [explosion_range]. Required radius was [required_range]. Core production complete."
+
+	playsound(
+		src,
+		pick('modular_zubbers/sound/arcade/minesweeper_win.ogg'),
+		50
+	)
+
+	return COMSIG_CANCEL_EXPLOSION
```
Comment (Majkl-J) [inline modular_zubbers/code/modules/alt_anomaly_refinery/anomaly_refinery.dm:117]: Ditto return change
```suggestion

	return
```

**Review by Majkl-J — CHANGES_REQUESTED** [review]

**Comment by StrangeWeirdKitten** [comment]
@BurgerLUA is this PR still applicable? Obviously it will need some changes to be more suited with the upstream anamoly changes. Primarily: limits on cores are gradually being introduced.

**Comment by BurgerLUA** [comment]
> @BurgerLUA is this PR still applicable? Obviously it will need some changes to be more suited with the upstream anamoly changes. Primarily: limits on cores are gradually being introduced.

It should still works, but tell me what changes you need done and I will do them to get this PR merged.

**Comment by StrangeWeirdKitten** [comment]
My biggest thing is that anamolies need to still respect the anamoly cap.

**Comment by StrangeWeirdKitten** [comment]
Forgot about this PR, but I'll test merge it.

File: modular_zubbers/code/modules/alt_anomaly_refinery/anomaly_refinery.dm [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+
+/obj/machinery/research/anomaly_refinery
+	desc = "An advanced machine equipped with state of the art bomb prediction software that's capable of implosion-compressing raw anomaly cores into finished artifacts. \
+	Comes with a built-in stabilization and sound-supressant field that consumes a lot of power to prevent obnoxious shaking and sounds heard station-wide, as not to wake up any dorms users."
+	var/requirement_timer
+	var/requirement_mod = 0 //0 to 100. Decreases by 1 every 5 seconds. Increased by 40 after a test.
+
+/obj/machinery/research/anomaly_refinery/examine(mob/user)
+	. = ..()
+	. += span_notice("Usage of the machine will increase the bomb range requirement for the next experiment, however this diminishes with time.")
```
Comment (StrangeWeirdKitten) [inline modular_zubbers/code/modules/alt_anomaly_refinery/anomaly_refinery.dm:12]:
```suggestion
	. += span_notice("Usage of the machine will increase the bomb range requirement for the next experiment.")
```

**Comment by pixelkitty286** [comment]
I still hate the fact it requires so much power in the grid which is 99% of the time never there meaning you can't ever refine.

**Comment by StrangeWeirdKitten** [comment]
It's time.

**Pull request merged**